File: app/src/main/java/com/robifr/ledger/ui/createqueue/viewmodel/SelectProductOrderViewModel.java
/**
 * Copyright (c) 2024 Robi
 *
 * Ledger is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Ledger is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Ledger. If not, see <https://www.gnu.org/licenses/>.
 */

package com.robifr.ledger.ui.createqueue.viewmodel;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import com.robifr.ledger.data.model.ProductOrderModel;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Objects;
import java.util.Set;
import java.util.TreeSet;
import java.util.stream.Collectors;

public class SelectProductOrderViewModel {
  @NonNull private final CreateQueueViewModel _viewModel;
  @NonNull private final MutableLiveData<Boolean> _isContextualModeActive = new MutableLiveData<>();

  /** Selected product order indexes from {@link CreateQueueViewModel#inputtedProductOrders()} */
  @NonNull private final MutableLiveData<Set<Integer>> _selectedIndexes = new MutableLiveData<>();

  public SelectProductOrderViewModel(@NonNull CreateQueueViewModel viewModel) {
    this._viewModel = Objects.requireNonNull(viewModel);
  }

  @NonNull
  public LiveData<Boolean> isContextualModeActive() {
    return this._isContextualModeActive;
  }

  @NonNull
  public LiveData<Set<Integer>> selectedIndexes() {
    return this._selectedIndexes;
  }

  public void onProductOrderCheckedChanged(int productOrderIndex, boolean isChecked) {
    final HashSet<Integer> selectedIndexes =
        this._selectedIndexes.getValue() != null
            ? new HashSet<>(this._selectedIndexes.getValue())
            : new HashSet<>();

    // Replace product image — its initial name — with checkbox
    // when being selected and vice versa.
    if (isChecked) selectedIndexes.add(productOrderIndex);
    else selectedIndexes.remove(productOrderIndex);

    if (selectedIndexes.isEmpty()
        // Prevent re-invoking the live data.
        && (this._isContextualModeActive.getValue() == null
            || this._isContextualModeActive.getValue())) {
      this._isContextualModeActive.setValue(false);

    } else if (!selectedIndexes.isEmpty()
        // Prevent re-invoking the live data.
        && (this._isContextualModeActive.getValue() == null
            || !this._isContextualModeActive.getValue())) {
      this._isContextualModeActive.setValue(true);
    }

    // Only set after contextual mode is set.
    // So that `ActionMode` instance already exist when we set its title.
    this._selectedIndexes.setValue(selectedIndexes);
  }

  public void onDeleteSelectedProductOrder() {
    final TreeSet<Integer> selectedIndexes =
        this._selectedIndexes.getValue() != null
            ? new TreeSet<>(this._selectedIndexes.getValue())
            : new TreeSet<>();
    final List<ProductOrderModel> productOrdersToDelete =
        selectedIndexes.stream()
            .map(index -> this._viewModel.inputtedProductOrders().get(index))
            .collect(Collectors.toList());

    this.reset();
    this._viewModel.onRemoveProductOrder(productOrdersToDelete.toArray(new ProductOrderModel[0]));
  }

  public void reset() {
    this._isContextualModeActive.setValue(false);
    this._selectedIndexes.setValue(Collections.unmodifiableSet(new HashSet<>()));
  }
}

File: app/src/main/java/com/robifr/ledger/ui/createqueue/viewmodel/SelectProductOrderViewModel.kt
/**
 * Copyright 2024 Robi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.robifr.ledger.ui.createqueue.viewmodel

import com.robifr.ledger.ui.SafeMutableLiveData

class SelectProductOrderViewModel(private val _viewModel: CreateQueueViewModel) {
  private val _uiState: SafeMutableLiveData<SelectProductOrderState> =
      SafeMutableLiveData(SelectProductOrderState(setOf()))
  val uiState: SafeMutableLiveData<SelectProductOrderState>
    get() = _uiState

  fun onProductOrderCheckedChanged(productOrderIndex: Int, isChecked: Boolean) {
    val selectedIndexes: MutableSet<Int> = _uiState.safeValue.selectedIndexes.toMutableSet()
    if (isChecked) selectedIndexes.add(productOrderIndex)
    else selectedIndexes.remove(productOrderIndex)

    _uiState.setValue(_uiState.safeValue.copy(selectedIndexes = selectedIndexes))
  }

  fun onDeleteSelectedProductOrder() {
    _viewModel.onProductOrdersChanged(
        _viewModel.uiState.safeValue.productOrders.toMutableList().apply {
          _uiState.safeValue.selectedIndexes.sortedDescending().forEach { removeAt(it) }
        })
    onDisableContextualMode()
  }

  fun onDisableContextualMode() {
    _uiState.setValue(SelectProductOrderState(setOf()))
  }
}
